framework/src/geomsearch/PenetrationLocator.C

```diff
@@ -121,6 +123,35 @@ PenetrationLocator::detectPenetration()
 
   Threads::parallel_reduce(slave_node_range, pt);
 
+  std::vector<dof_id_type> recheck_slave_nodes = pt._recheck_slave_nodes;
+
+  // Update the patch for the slave nodes in recheck_slave_nodes and re-run penetration thread on
+  // these nodes at every nonlinear iteration if patch update strategy is set to "nonlinear_iter"
+  // which is 3.
+  if (recheck_slave_nodes.size() > 0 && _patch_update_strategy == 3 &&
+      _subproblem.currentlyComputingJacobian())
+  {
+    // Update the patch for this subset of slave nodes and calculate the nearest neighbor_nodes
+    _nearest_node.updatePatch(recheck_slave_nodes);
+
+    // Re-run the penetration thread to see if these nodes are in contact with the updated patch
+    NodeIdRange recheck_slave_node_range(recheck_slave_nodes.begin(), recheck_slave_nodes.end(), 1);
+
+    Threads::parallel_reduce(recheck_slave_node_range, pt);
+  }
+
+  if (recheck_slave_nodes.size() > 0 && _patch_update_strategy != 3 &&
+      _subproblem.currentlyComputingJacobian() && _first_warning)
+  {
+    mooseWarning("Warning in PenetrationLocator. Penetration is not detected for some slave nodes. "
+                 "If all slave nodes are expected to be in contact with the master boundary, "
+                 "please set 'patch_update_strategy' to 'nonlinear_iter' in the Mesh block.");
```

Comment:
I'd change the warning message to:

Penetration is not detected for one or more slave nodes. This could be because those slave nodes simply do not project to faces on the master surface. However, this could also be because contact should be enforced on those nodes, but the faces that they project to are outside the contact patch, which will give an erroneous result. Use appropriate options for 'patch_size' and 'patch_update_strategy' in the Mesh block to avoid this issue. Setting 'patch_update_strategy=nonlinear_iter' is recommended because it completely avoids this potential issue.

Reply:
I have changed the warning message to the one above.